File: ios/SYCCourses/Sources/SYCCourses/Licensing/LicensingAPIClient.swift
import Foundation

struct ActivationResponse: Decodable, Sendable { let entitlement: EntitlementEnvelope; let refreshCredential: String }
struct RefreshResponse: Decodable, Sendable { let entitlement: EntitlementEnvelope }

enum LicensingAPIError: Error, LocalizedError {
    case invalidResponse, server(String)
    var errorDescription: String? { switch self { case .invalidResponse: "The licensing service returned an invalid response."; case .server(let code): code } }
}

struct LicensingAPIClient: Sendable {
    let endpoint: URL
    let session: URLSession

    func activate(code: String, installationId: String, appVersion: String, platform: String) async throws -> ActivationResponse {
        try await post("v1/activations", body: ["invitationCode": code, "installationId": installationId, "appVersion": appVersion, "platform": platform])
    }

    func refresh(credential: String, installationId: String, appVersion: String) async throws -> RefreshResponse {
        try await post("v1/entitlements/refresh", body: ["refreshCredential": credential, "installationId": installationId, "appVersion": appVersion])
    }

    private func post<T: Decodable>(_ path: String, body: [String: String]) async throws -> T {
        var request = URLRequest(url: endpoint.appendingPathComponent(path))
        request.httpMethod = "POST"; request.timeoutInterval = 15
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(body)
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw LicensingAPIError.invalidResponse }
        guard (200..<300).contains(http.statusCode) else {
            let code = (try? JSONSerialization.jsonObject(with: data) as? [String: Any]).flatMap { $0["error"] as? [String: Any] }?["code"] as? String
            throw LicensingAPIError.server(code ?? "TEMPORARY_SERVER_ERROR")
        }
        do { return try JSONDecoder().decode(T.self, from: data) } catch { throw LicensingAPIError.invalidResponse }
    }
}

File: android/app/src/main/java/com/lukematthews/syccourses/LicensingApiClient.kt
package com.lukematthews.syccourses

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.net.HttpURLConnection
import java.net.URL

@Serializable private data class ActivationRequest(val invitationCode: String, val installationId: String, val appVersion: String, val platform: String = "android")
@Serializable private data class RefreshRequest(val refreshCredential: String, val installationId: String, val appVersion: String)
@Serializable data class ActivationResponse(val entitlement: EntitlementEnvelope, val refreshCredential: String)
@Serializable data class RefreshResponse(val entitlement: EntitlementEnvelope)

class LicensingApiException(val code: String) : Exception(code)

class LicensingApiClient(private val endpoint: String, private val json: Json = Json { ignoreUnknownKeys = true }) {
    suspend fun activate(code: String, installationId: String, appVersion: String): ActivationResponse =
        post("v1/activations", json.encodeToString(ActivationRequest(code, installationId, appVersion)))

    suspend fun refresh(credential: String, installationId: String, appVersion: String): RefreshResponse =
        post("v1/entitlements/refresh", json.encodeToString(RefreshRequest(credential, installationId, appVersion)))

    private suspend inline fun <reified T> post(path: String, body: String): T = withContext(Dispatchers.IO) {
        val connection = URL("${endpoint.trimEnd('/')}/$path").openConnection() as HttpURLConnection
        try {
            connection.requestMethod = "POST"
            connection.connectTimeout = 15_000
            connection.readTimeout = 15_000
            connection.doOutput = true
            connection.setRequestProperty("Content-Type", "application/json")
            connection.outputStream.use { it.write(body.encodeToByteArray()) }
            val stream = if (connection.responseCode in 200..299) connection.inputStream else connection.errorStream
            val responseBody = stream?.use { it.readBytes().decodeToString() }.orEmpty()
            if (connection.responseCode !in 200..299) {
                val code = runCatching { json.parseToJsonElement(responseBody).jsonObject["error"]!!.jsonObject["code"]!!.jsonPrimitive.content }.getOrNull()
                throw LicensingApiException(code ?: "TEMPORARY_SERVER_ERROR")
            }
            runCatching { json.decodeFromString<T>(responseBody) }.getOrElse { throw LicensingApiException("INVALID_RESPONSE") }
        } finally { connection.disconnect() }
    }
}
